### PR TITLE
[IMM32] Rewrite ImmRegisterWordA

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose

Implementing Japanese input...
JIRA issue: [CORE-11700](https://jira.reactos.org/browse/CORE-11700)

## Proposed changes

- Rewrite `ImmRegisterWordA` function.
- Modify `imetable.h`.

### CODE REVIEW

File: dll/win32/imm32/imm.c [hunk 1]
```diff
@@ -416,7 +416,7 @@ static PIMEDPI APIENTRY Ime32LoadImeDpi(HKL hKL, BOOL bLock)
     }
 }
 
-BOOL WINAPI ImmLoadIME(HKL hKL)
+PIMEDPI WINAPI ImmLoadIME(HKL hKL)
```
Comment (HBelusca) [inline dll/win32/imm32/imm.c:419]: No, the function returns a boolean.

File: dll/win32/imm32/imm.c [hunk 2]
```diff
@@ -3184,28 +3184,56 @@ BOOL WINAPI ImmNotifyIME(
...
+    INT cch;
+
+    TRACE("(%p, %s, 0x%lX, %s)\n", hKL, debugstr_a(lpszReading), dwStyle,
+          debugstr_a(lpszRegister));
+
+    pImeDpi = ImmLoadIME(hKL);
+    if (!pImeDpi)
+        return FALSE;
+
+    if (!(pImeDpi->ImeInfo.fdwProperty & IME_PROP_UNICODE))
```
Comment (HBelusca) [inline dll/win32/imm32/imm.c:3199]: You retrieve the pImeDpi via a call to `ImmLockImeDpi(hKL)`.

**Review by HBelusca — CHANGES_REQUESTED** [review]

File: dll/win32/imm32/imm.c (hunk 1 above)
Comment (katahiromz) [reply dll/win32/imm32/imm.c]: Fixed.

File: dll/win32/imm32/imm.c (hunk 2 above)
Comment (katahiromz) [reply dll/win32/imm32/imm.c:3220]: OK.

File: dll/win32/imm32/imm.c [hunk 3]
```diff
@@ -437,6 +437,27 @@ BOOL WINAPI ImmLoadIME(HKL hKL)
...
 
+PIMEDPI APIENTRY ImmLockOrLoadImeDpi(HKL hKL)
+{
+    PW32CLIENTINFO pInfo;
+    PIMEDPI pImeDpi;
+
+    if (!IS_IME_HKL(hKL))
+    {
+        if (!g_psi || !(g_psi->dwSRVIFlags & SRVINFO_CICERO_ENABLED))
+            return FALSE;
```
[the remark this answers is not in the document]
Comment (katahiromz) [reply dll/win32/imm32/imm.c]: OK.

File: dll/win32/imm32/imm.c [hunk 4]
```diff
@@ -437,6 +437,27 @@ BOOL WINAPI ImmLoadIME(HKL hKL)
...
+
+        pInfo = (PW32CLIENTINFO)(NtCurrentTeb()->Win32ClientInfo);
+        if ((pInfo->W32ClientInfo[0] & 2))
+            return FALSE;
```
[the remark this answers is not in the document]
Comment (katahiromz) [reply dll/win32/imm32/imm.c]: OK.

**Comment by katahiromz** [comment]
Check time. I will merge 4 days later.

**Comment by katahiromz** [comment]
No response?

**Pull request merged**